src/pow.rs

```diff
@@ -0,0 +1,1687 @@
+// Rust Bitcoin Library - Written by the rust-bitcoin developers.
+// SPDX-License-Identifier: CC0-1.0
+
+//! Proof-of-work related integer types.
+//!
+//! Provides the [`Work`] and [`Target`] types that are use in proof-of-work calculations. The
+//! functions here are designed to be fast.
+//!
+
+use core::convert::TryFrom;
+#[cfg(feature = "serde")]
+use core::convert::TryInto;
+use core::fmt::{self, LowerHex, UpperHex};
+use core::ops::{Add, Div, Mul, Not, Rem, Shl, Shr, Sub};
+
+use crate::blockdata::constants;
+#[cfg(doc)]
+use crate::consensus::Params;
+
+/// Implements `From<$int>` for `$ty` where `$ty` is an unsigned integer type.
+macro_rules! impl_from_unsigned_type {
+    ($ty:ident, $int:ident) => {
+        impl From<$int> for $ty {
+            #[inline]
+            fn from(x: $int) -> $ty { $ty(Uint256::from(x)) }
+        }
+    };
+}
+
+/// Implements `TryFrom<$int>` for `$ty` where `$ty` is a signed integer type.
+macro_rules! impl_try_from_signed_type {
+    ($ty:ident, $int:ident) => {
+        impl TryFrom<$int> for $ty {
+            type Error = TryFromError;
+            #[inline]
+            fn try_from(x: $int) -> Result<Self, Self::Error> {
+                let x = Uint256::try_from(x)?;
+                Ok($ty(x))
+            }
+        }
+    };
+}
+
+// Implement traits and methods shared by `Target` and `Work`.
+macro_rules! do_impl {
+    ($ty:ident) => {
+        impl $ty {
+            #[doc = concat!("Creates a [`", stringify!($ty), "`] from a big-endian byte array.")]
+            #[inline]
+            pub fn from_be_bytes(bytes: [u8; 32]) -> $ty { $ty(Uint256::from_be_bytes((bytes))) }
+
+            #[doc = concat!("Creates a [`", stringify!($ty), "`] from a little-endian byte array.")]
+            #[inline]
+            pub fn from_le_bytes(bytes: [u8; 32]) -> $ty { $ty(Uint256::from_le_bytes((bytes))) }
+
+            /// Converts `self` to a big-endian byte array.
+            #[inline]
+            pub fn to_be_bytes(self) -> [u8; 32] { self.0.to_be_bytes() }
+
+            /// Converts `self` to a little-endian byte array.
+            #[inline]
+            pub fn to_le_bytes(self) -> [u8; 32] { self.0.to_le_bytes() }
+        }
+
+        impl_from_unsigned_type!($ty, u8);
+        impl_from_unsigned_type!($ty, u16);
+        impl_from_unsigned_type!($ty, u32);
+        impl_from_unsigned_type!($ty, u64);
+        impl_from_unsigned_type!($ty, u128);
+
+        impl_try_from_signed_type!($ty, i8);
+        impl_try_from_signed_type!($ty, i16);
+        impl_try_from_signed_type!($ty, i32);
+        impl_try_from_signed_type!($ty, i64);
+        impl_try_from_signed_type!($ty, i128);
+
+        impl From<[u8; 32]> for $ty {
+            #[doc = concat!("Creates a [`", stringify!($ty), "`] from a big-endian array of bytes.")]
+            #[inline]
+            fn from(a: [u8; 32]) -> Self { $ty(Uint256::from(a)) }
+        }
```

Comment:
Why did you write `From<{array}>` instead of inherent `from_{endianess} function? The endianess ambiguity footgun is pretty serious. If you want to avoid duplication:

```rust
pub trait FromArray<A>: sealed::FromArray where A: AsMut<[u8]> {
    fn from_be(array: A) -> Self;

    fn from_le(mut array: A) -> Self {
        array.as_mut().reverse();
        Self::from_be(array)
    }
}
```

And probably have inherent functions so that people don't have to bother importing the trait for something that silly.

Reply:
>The endianess ambiguity footgun is pretty serious.

This made me laugh, in a somewhat dark way .... Satoshi was not careful about this and the consequence 13 years later is that we're still fighting about why half the hashes in Bitcoin are reversed half the time. So the consequences could be pretty bad, indeed.

Reply:
I've added extensive constuctors of form `from_be_array_u8` (u8, u16, u32, u64, u128 - le and be) and the same for the other direction (e.g. `to_be_array_u8`). Also added convenience methods `to/from_be_bytes` (u8) and `to/from_be` (u128). 

All roundtrip unit tested.

Reply:
I wasn't able to decipher your code above @Kixunil, I've used macros instead.

Reply:
I'm not enamored with the names, open to suggestions please.

Reply:
The trait was supposed to solve the naming - the inherent function could be named `from_le` and `from_be`.

Maybe more code helps:

`pub fn from_le<A>(array: A) -> Self where Self: FromArray<A>`

`impl `FromArray<[{unsigned}; N]> for Target { /* ... */ }`

Reply:
Aah, this is the same trick we did in `LockTime::is_satisfied_by(t: T)`, amusingly this realisation came to me mid way through a show at the theater on the weekend :)